cpp/tests/utility/vector_equality.hpp

```diff
@@ -147,6 +162,19 @@ inline void expect_vector_equivalent(Vector1 const& lhs, Vector2 const& rhs)
     EXPECT_THAT(to_host<T>(lhs), ::testing::Pointwise(float_matcher(), to_host<T>(rhs)));
   } else if constexpr (std::is_integral_v<T>) {
     EXPECT_THAT(to_host<T>(lhs), ::testing::Pointwise(::testing::Eq(), to_host<T>(rhs)));
+  } else if constexpr (cuspatial::is_optional<T>::value) {
+    if constexpr (cuspatial::is_vec_2d<typename T::value_type>()) {
+      EXPECT_THAT(to_host<T>(lhs),
+                  ::testing::Pointwise(optional_matcher(vec_2d_matcher()), to_host<T>(rhs)));
+    } else if constexpr (std::is_floating_point_v<typename T::value_type>) {
+      EXPECT_THAT(to_host<T>(lhs),
+                  ::testing::Pointwise(optional_matcher(float_matcher()), to_host<T>(rhs)));
+    } else if constexpr (std::is_integral_v<typename T::value_type>) {
+      EXPECT_THAT(to_host<T>(lhs),
+                  ::testing::Pointwise(optional_matcher(::testing::Eq()), to_host<T>(rhs)));
+    } else {
+      EXPECT_EQ(lhs, rhs);
+    }
```

Comment:
This part is a bit unfortunate.. Maybe it prompts us to do some refactoring of the vector equality utility.